lib/react_client.dart

```diff
@@ -93,61 +88,67 @@ class ReactDartComponentFactoryProxy extends ReactComponentFactoryProxy {
       Map props = invocation.positionalArguments[0];
       List children = invocation.positionalArguments.sublist(1);
 
-      List reactParams = [generateExtendedJsProps(props, children)];
-      reactParams.addAll(children);
+      markChildrenValidated(children);
 
-      return reactComponentFactory.apply(reactParams);
+      return reactComponentFactory(
+        generateExtendedJsProps(props, children, defaultProps: defaultProps),
+        jsifyChildren(children)
+      );
     }
 
     return super.noSuchMethod(invocation);
   }
 
-  /// Returns a [JsObject] version of the specified [props], preprocessed for consumption by ReactJS and prepared for
+  /// Returns a JavaScript version of the specified [props], preprocessed for consumption by ReactJS and prepared for
   /// consumption by the [react] library internals.
-  static JsObject generateExtendedJsProps(Map props, dynamic children) {
+  static InteropProps generateExtendedJsProps(Map props, dynamic children, {Map defaultProps}) {
     if (children == null) {
       children = [];
     } else if (children is! Iterable) {
       children = [children];
     }
 
-    Map extendedProps = new Map.from(props);
-    extendedProps['children'] = children;
+    // 1. Merge in defaults (if they were specified)
+    // 2. Add specified props and children.
+    // 3. Remove "reserved" props that should not be visible to the rendered component.
 
-    JsObject jsProps = newJsObjectEmpty();
-
-    // Transfer over `key` if specified so ReactJS knows about it.
-    if (extendedProps.containsKey('key')) {
-      jsProps['key'] = extendedProps['key'];
+    // [1]
+    Map extendedProps = (defaultProps != null ? new Map.from(defaultProps) : {})
+      // [2]
+      ..addAll(props)
+      ..['children'] = children
+      // [3]
+      ..remove('key')
+      ..remove('ref');
+
+    var internal = new ReactDartComponentInternal()
+      ..props = extendedProps;
+
+    var interopProps = new InteropProps(internal: internal);
+
+    // Don't pass a key into InteropProps if one isn't defined, so that the value will
+    // be `undefined` in the JS, which is ignored by React, whereas `null` isn't.
+    if (props.containsKey('key')) {
+      interopProps.key = props['key'];
     }
 
-    // Transfer over `ref` if specified so ReactJS knows about it.
-    if (extendedProps.containsKey('ref')) {
-      var ref = extendedProps['ref'];
+    if (props.containsKey('ref')) {
+      var ref = props['ref'];
 
       // If the ref is a callback, pass ReactJS a function that will call it
-      // with the Dart Component instance, not the JsObject instance.
+      // with the Dart Component instance, not the ReactComponent instance.
       if (ref is _CallbackRef) {
-        jsProps['ref'] = (JsObject instance) => ref(instance == null ? null : _getComponent(instance));
+        interopProps.ref = allowInterop((ReactComponent instance) =>
+            ref(instance == null ? null : instance.props.internal.component));
```

Comment:
are you intentionally avoiding using null aware ops for backward compatibility sake?